workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/SpatialOSBehaviourManager.cs

```diff
@@ -1,60 +1,202 @@
-            // Setup caches
+            behaviourLibrary = library;
+
+            var spatialComponent = gameObject.GetComponent<SpatialOSComponent>();
+            entity = spatialComponent.Entity;
+            spatialId = spatialComponent.SpatialEntityId;
+
+            foreach (var behaviour in gameObject.GetComponents<MonoBehaviour>())
+            {
+                var readerIds = library.GetRequiredReaderComponentIds(behaviour.GetType());
+                foreach (var id in readerIds)
+                {
+                    if (!componentIdToReaderSpatialOSBehaviours.ContainsKey(id))
+                    {
+                        componentIdToReaderSpatialOSBehaviours.Add(id, new HashSet<MonoBehaviour>());
+                    }
+
+                    componentIdToReaderSpatialOSBehaviours[id].Add(behaviour);
+                }
+
+                var writerIds = library.GetRequiredWriterComponentIds(behaviour.GetType());
+                foreach (var id in writerIds)
+                {
+                    if (!componentIdToWriterSpatialOSBehaviours.ContainsKey(id))
+                    {
+                        componentIdToWriterSpatialOSBehaviours.Add(id, new HashSet<MonoBehaviour>());
+                    }
+
+                    componentIdToWriterSpatialOSBehaviours[id].Add(behaviour);
+                }
+
+                numUnsatisfiedReadersOrWriters[behaviour] = readerIds.Count + writerIds.Count;
+            }
         }
 
-        public List<ReaderWriterImpl> GetReadersWriters(uint componentId)
+
+        public HashSet<IReaderInternal> GetReadersWriters(uint componentId)
         {
-            return null;
+            return componentIdToCreatedReadersWriters[componentId];
         }
 
         public void EnableSpatialOSBehaviours()
         {
-            // inject + enable all spatialOSBehavioursToEnable
+            foreach (var behaviour in spatialOSBehavioursToEnable)
+            {
+                var dict = behaviourLibrary.InjectAllReadersWriters(behaviour, entity);
+                behaviourToCreatedReadersWriters[behaviour] = dict;
+                foreach (var idToReaderWriter in dict)
+                {
+                    var id = idToReaderWriter.Key;
+                    var reader = idToReaderWriter.Value;
+                    if (!componentIdToCreatedReadersWriters.ContainsKey(id))
+                    {
+                        componentIdToCreatedReadersWriters[id] = new HashSet<IReaderInternal>();
+                    }
+
+                    componentIdToCreatedReadersWriters[id].Add(reader);
+                }
+
+                behaviour.enabled = true;
+            }
+
+            spatialOSBehavioursToEnable.Clear();
         }
 
         public void DisableSpatialOSBehaviours()
         {
-            // disable + deinject all spatialOSBehavioursToDisable
+            foreach (var behaviour in spatialOSBehavioursToDisable)
+            {
+                behaviour.enabled = false;
```

Comment:
User `OnDisable` code could throw an exception - double-check that it's not propagated up to this level or you'll need to protect against it.